src/local.py

```diff
@@ -1,84 +1,113 @@
+    IS_WINDOWS,
+)
+import utils
+
+
+if IS_WINDOWS:
     import winreg
 
-from consts import MINECRAFT_REGISTRY_PATHS, REGISTRY_START_PATHS, MINECRAFT_REGISTRY_PATH_INSTALL_LOCATION_KEY, WINDOWS_UNINSTALL_LOCATION
+log = logging.getLogger(__name__)
 
-import logging as log
-import os
-import psutil
-import asyncio
 
-class LocalClient():
+class LocalClient:
     def __init__(self):
-        self.running_process = None
-
-    async def get_size_at_path(self, start_path):
-        total_size = 0
-        for dirpath, dirnames, filenames in os.walk(start_path):
-            for f in filenames:
-                await asyncio.sleep(0)
-                fp = os.path.join(dirpath, f)
-                if not os.path.islink(fp):
-                    total_size += os.path.getsize(fp)
-
-        return total_size
-
-    def get_minecraft_launcher_path(self):
-        if sys.platform == 'win32':
-            install_path = self.get_minecraft_folder_path()
-            potential_path = os.path.join(install_path, 'MinecraftLauncher.exe')
-            if os.path.exists(potential_path):
-                return potential_path
-        else:
-            return self.get_minecraft_folder_path()
-
-    def get_minecraft_folder_path(self):
-        if sys.platform == 'win32':
-            for start_path in REGISTRY_START_PATHS:
-                for minecraft_reg_path in MINECRAFT_REGISTRY_PATHS:
-                    try:
-                        reg = winreg.ConnectRegistry(None, start_path)
-                        with winreg.OpenKey(reg, minecraft_reg_path) as key:
-                            install_path = winreg.QueryValueEx(key, MINECRAFT_REGISTRY_PATH_INSTALL_LOCATION_KEY)[0]
-                    except OSError:
-                        continue
-                    if os.path.exists(install_path):
-                        return install_path
-        else:
-            potential_path = "/Applications/Minecraft.app"
-            if os.path.exists(potential_path):
-                return potential_path
+        self.running_games = {
+            GameID.Minecraft: None,
+            GameID.MinecraftDungeons: None,
+        }
+
+    def find_launcher_path(self, game_id, *, folder=False, folder_path=None, exe=None) -> str:
+        if exe is None and folder_path is not None:
+            folder_path = None
+        if exe is not None and folder_path is None:
+            folder_path = exe
+        if exe is not None and folder_path is not None:
+            folder_path = os.path.abspath(folder_path)
+            exe = os.path.abspath(exe)
+        log.debug(f"folder path for {game_id}: {folder_path}")
+        log.debug(f"path for {game_id}: {exe}")
+        return exe if not folder else folder_path
+
+    def is_game_still_running(self, game_id) -> bool:
+        rp = self.running_games[game_id]  # Just an alias
+        if rp is not None and rp.poll() is None:
+            return True
+        elif rp is not None and not rp.poll() is not None:

```

Comment:
Sorry, I'm not sure what you're talking about here. But I'll try to answer as best as I can. Looking at it I realise I could've written together but I'm not sure where the bug is. The `rp is not None` line checks if `self.running_games[game_id]` is None and thus has been assigned a process at some point (and yes, I could've done this in one if statement wrapping the two next statements). `rp.poll()` is just to see if the process returned when we launched the game is still running. I used this instead of psutils so we don't have to repeatedly iterate over a list of processes and try to find the game.

Reply:
i think this is a bit confusing ``and not rp.poll() is not None:`` shouldnt it be just ``and rp.poll() is not None:`` instead ?

Reply:
TouwaStar is right, `not poll()` is bool so it is always `not None`

- process is still running:
`poll() == None` -> `not poll() == True` -> `not poll() is not None == True`
- process stopped with 0 status code:
`poll() == 0` -> `not poll() == True` -> `not poll() is not None == True`
- process stopped with status code other than 0:
`poll() == 1` -> `not poll() == False` -> `not poll() is not None == True`

Reply:
Oh, I see the issue now 🤦. I think I can refactor it to something like this:
```python
def is_game_still_running(self, game_id) -> bool:
	return (self.running_games[game_id] is not None 
	        and self.running_games[game_id] is None)
```
There's no point of reassigning None as another process will just replace whatever the value is of the property `self.running_games[game_id]` when a game starts. I can simplify the code down to this basically.

Reply:
yes, almost - you forget inspect `poll()`

```python
def is_game_still_running(self, game_id) -> bool:
    return self.running_games[game_id] and self.running_games[game_id].poll() is None
```

Reply:
🤦